Simon iterative compression

SDaneau opened the pull request:
Many things to verify:

1. First, you need to define a beam by using the vertical calibration tab. You can define LO (you can change the name) between 100 and 200. You can also create a second one "A" between 400 and 500. Click on "assign beam and period".
2. Check that the beam name list in temporal calibration tab is updated to the beam you just created. 
3. In demo mode, acquire the dummy temporal data. You can apply an SNR threshold of 2.0. You can select the wavelength bandwidth between 200 and 300. Click on "fit chirp scan" and "assign calibration". The parameters will be updated in the optimal phase of the selected beam name in the beam explorer. The name selected in the temporal calibration tab define for which beam the phase is assigned.
4. If you click a second time on "assign calibration" you'll get twice the values in the optimal phase. This action mimic many chirp scan fit to optimize the phase. 
5. Let's say you did it wrong and you want to start over the optimization, you can click on "clear optimal" in the beam explorer. This will put the optimal phase parameters to 0. 
6. So starting again, click on "assign calibration" to get some optimal values. Now we want to send different phases to the SLM.
7. In absolute mode (unchecked the relative box in the beam explorer), if you click on "apply beams" then you should only get the grating. Changing the values in "current phase" and clicking again on "apply beams" gives you something else. 
8. In relative mode (checked the relative box), the values in "current phase" will be added to the ones in "optimal phase". A good test is to put the inverse values so the sum gives 0 for all coefficients. Do such and by clicking on "apply beams", it gives you the grating. Otherwise, for different current phase values, you can play around the optimal values. 

TO BE DONE BEFORE MERGING: 

- Try with real chirp scan fit to see if we can compress the light as it should. It will give us some hits relatively to the SLM grayscale control and also wavelength calibration.

SDaneau committed:
Merge branch 'beam_explorer' into Simon_IterativeCompression

SDaneau committed:
Merged beam_explorer in the branch

Resolved issues

SDaneau committed:
Solved issues

Error on how SLM device is called. The Beam explorer window is opening now,

SDaneau committed:
Update main.py

SDaneau committed:
SNR Threshold

Change it to a Double spin box

SDaneau committed:
Update beam name

In temporal calibration

SDaneau committed:
Implement the assign chirp coeff

Send the coeff to the beam widget, adjust the number of column.

SDaneau committed:
Update the beam explorer

SDaneau committed:
Merge branch 'dev' into Simon_IterativeCompression

SDaneau committed:
Bugs in communication with beam explorer

SDaneau committed:
Update beams.py

SDaneau committed:
Clear optimal coefficient

SDaneau committed:
Update beam explorer functionalities

Making sure all the functions are working well and we can play with the SLM by applying any phase, relative or absolute, from an optimal fitted phase.

fthouin committed:
Added relative path to load demo dataset

fthouin commented:
Thanks for the contribution Simon!

1. I added a line so that `Chirp_dataset.txt` is found by the code irrespective of the current working directory (e.g. running python main.py from /src/ or running python /src/main.py from / )
2. Everything described seems to work fine. However, if I want to manually change the optimal phase by changing the coefficients in the table and pressing Apply Beams, the optimal phase coefficients are reset to the previous value. I don't think that is the expected behavior.
3. Are the phase plot ok?
<img width="620" height="300" alt="image" src="https://github.com/user-attachments/assets/865e091b-f5cf-483e-b603-e2606141d805" />
In the ccase above, I would have expected a flat zero relative phase when Plot relative is checked. Any idea what is going on?

fthouin commented on src/measurements/CalibrationClasses.py:
```diff
@@ -384,27 +379,28 @@ def run(self):
-        np.savetxt('chirp.txt', self.chirp_)
+        np.savetxt('chirp.txt', self.chirp)
         np.savetxt('wls.txt', self.wls)
         np.savetxt('intensities.txt', np.array(self.intensities))
```
Is this for testing? If there is a need to save data, then it might be time to do it like we would in the final version of the code (H5 files and all)

SDaneau committed:
Manually change the optimal phase

SDaneau committed:
Merge branch 'Simon_IterativeCompression' of https://github.com/SilvaScience/colberto into Simon_IterativeCompression

SDaneau commented:
Did the vertical calibration and the wavelength calibration. The second screenshot is showing the first chirp scan.

<img width="1413" height="1034" alt="Screenshot 2025-10-08 124241" src="https://github.com/user-attachments/assets/fb45273a-26ca-4009-86df-af29c1ea053b" />

<img width="1413" height="1034" alt="Screenshot 2025-10-08 124039" src="https://github.com/user-attachments/assets/b7d948bd-5394-471a-8193-81c9d34dcf5e" />

By applying the chirp parameters and doing the second chirp scan, we get signal now located at higher chirp value and more spread. The general shape is looking the same tho. 

<img width="1413" height="1034" alt="Screenshot 2025-10-08 125222" src="https://github.com/user-attachments/assets/60295a94-9876-4d7d-bafd-ecdc355ef297" />

SDaneau commented:
Without and with a UVFS rod (thickness of 2cm, GVD = 350fs2/cm at 800nm). Clearly, the chirp scan isn't good. 
<img width="1413" height="1034" alt="Screenshot 2025-10-08 150004" src="https://github.com/user-attachments/assets/ea278aaa-e4a0-45b5-bcc7-e00e9735f198" />
<img width="1413" height="1034" alt="Screenshot 2025-10-08 145711" src="https://github.com/user-attachments/assets/98f4c5a7-976d-43ec-800c-0dc7a156d31f" />

SDaneau commented:
Other example after correcting for the wavelength calibration... the same problem appears seems the chirp fit good but we can't correct the chirp accordingly.

<img width="3840" height="1200" alt="Screenshot 2025-10-09 154355" src="https://github.com/user-attachments/assets/463b432b-c461-4756-88ae-44154c393a03" />
<img width="3840" height="1200" alt="Screenshot 2025-10-09 154516" src="https://github.com/user-attachments/assets/d88dd2d8-21e7-4440-b4e0-f1e28f0913cf" />

SDaneau committed:
Major modifications

- Adjust the number of pixel of the stresing
- Modify the config_file for the calibration
- Now playing with the 785nm LUT file

SDaneau committed:
Add comments (personal notes)

SDaneau committed:
Right spectral calibration

SDaneau committed:
It works!

Changed the unit of the spectral calibration to be in meter

SDaneau committed:
Merge branch 'Simon_IterativeCompression' of https://github.com/SilvaScience/colberto into Simon_IterativeCompression

SDaneau committed:
Removed prints

SDaneau committed:
Final modifications

SDaneau commented:
Did the piece of glass test to increase the GDD. The change of GDD for 2cm of UVFS should be around 700fs^2. I obtained a change of roughly 500 fs^2 which is close but still different from what is expected. I've put the glass piece before the SLM, just after the DOE. Maybe it's too far from the SLM, so it also change a bit the beam path. I'll bring the UVFS glass close to the SHG crystal so it's gonna minimize the change in beam aligment.

SDaneau commented:
@fthouin anyway, I think you can do the review of the pull request. As far as I test it, I think we can push it to the dev branch. I'll follow up with the little test result.

fthouin commented on src/drivers/SLM.py:
```diff
@@ -206,7 +206,8 @@ def run(self):
             # 1) Connect to the SDK
             self.slm = self.create_slm_sdk()
-            self.load_lut(r"C:\Program Files\Meadowlark Optics\Blink 1920 HDMI\LUT Files\19x12_8bit_linearVoltage.lut")
+            #self.load_lut(r"C:\Program Files\Meadowlark Optics\Blink 1920 HDMI\LUT Files\19x12_8bit_linearVoltage.lut")
```
Why did you change the LUT file called here? What were the consequence on the experiment?

SDaneau replied:
I left the last LUT file called in comment. We were using a general 8 bit linear voltage. Now, I'm using the calibration for 785nm. This clearly changes the phase applied as the two calibration are slightly different for our wavelengths. However, I did not face any consequences.

fthouin replied:
Consequence don't need to be negative but it would be nice to know what happens when you change that. Especially when you use the SLM at longer wavelengths that the nominal one beyond which a full 2pi modulation is no longer possible.

fthouin replied:
We'll find a way to manage LUT files later for now it is ok

fthouin commented on src/drivers/Stresing.py:
```diff
@@ -48,7 +48,7 @@ def __init__(self,hardware_params):
         path_dll = str(path_dll)
 
-        path_config = Path(r"C:\Program Files\Stresing\Escam\config.ini")
+        path_config = Path(r"C:\Program Files\Stresing\Escam\config_UdeM.ini")
```
Why did you change the config file name instead of locally changing the config file contents?

SDaneau replied:
The config.ini file is the one from Stresing. I don't want to touch it. Instead, I created a new one containing only the useful information.

fthouin replied:
Ok it can work like that for now but we need to harmonize the configuration of the experiment. We can [do that later](https://openproject.silvascience.org/wp/895)

fthouin commented on src/drivers/Stresing.py:
```diff
@@ -301,7 +301,7 @@ def calculate_wavelength_array(self):
                 # 546.07        611.90
                 # 1013.98       1111.80
-                self.wavelengths = 0.9402*self.wavelengths-30.864
+                self.wavelengths = 0.955*self.wavelengths-30.864
```
These values should not be harded-coded in the driver but rather provided at initialization

SDaneau replied:
Done, parameters are in the config_UdeM file

SDaneau replied:
I've put the calibrations parameters in instrument.py instead of the calibration file.

fthouin commented on src/drivers/StresingDriver.py:
```diff
@@ -99,12 +99,12 @@ def init_driver(self, path_dll, path_config):
-    self.settings.camera_settings[self.drvno].VFREQ = int(config.get("board0","vfreq")) # Controls the vertical clock frequency for FFT sensors (sensor S14290). 
-    self.settings.camera_settings[self.drvno].use_software_polling = int(config.get("board0","useSoftwarePolling")) # Determines which method is used to copy data from DMA to user buffer.
-    self.settings.camera_settings[self.drvno].adc_gain = int(config.get("board0","adcGain")) # Controlling the gain function of the ADC in 3030 high speed cameras (sensor S14290 use 5 or 6).
+    self.settings.camera_settings[self.drvno].CAMCNT = int(config.get("Board0","camcnt")) # Number of cameras which are connected to one PCIe board.
```
Why change the names of these attributes. Did the nomenclature of the config file change?

SDaneau replied:
Because in the config file it's cap B instead of b.

fthouin replied:
@KatieKoch2 can you confirm the config file at WakeForest also share this nomenclature?

KatieKoch2 replied:
Our configuration file has lowercase board naming, so they are not the same

fthouin replied:
Thanks for letting us know. Simon, can you find a simple way to load the configuration that is case-insensitive?

SDaneau replied:
Done, replaced Board0 by board0 in the config file and in the script.

SDaneau replied:
I added the following third line to the script: 

config = configparser.ConfigParser()
config.read("config.ini")
config._sections = {k.lower(): v for k, v in config._sections.items()} # Lowercase all section names

SDaneau replied:
I added the following class to stresing.py to be case-insensitive: 

class CaseInsensitiveConfig(configparser.ConfigParser):
    
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.optionxform = str.lower  # lowercase all option keys

    def read(self, filenames, encoding=None):
        super().read(filenames, encoding)
        # lowercase section names
        self._sections = {k.lower(): {kk.lower(): vv for kk, vv in v.items()}
                          for k, v in self._sections.items()}

    # override .get() to lowercase section/option lookups
    def get(self, section, option, **kwargs):
        return super().get(section.lower(), option.lower(), **kwargs) 


For example all of these will now work the same:
v1 = config.get("Board0", "DacCameraChannel3")
v2 = config.get("board0", "daccamerachannel3")
v3 = config.get("BOARD0", "DACCAMERACHANNEL3")

fthouin replied:
So it seems like that code ChatGPT wrote only configures the package configParser. Reading the [documentation of configParser](https://docs.python.org/3/library/configparser.html), it states that "_Note also that keys in sections are case-insensitive and stored in lowercase_".

Bottom line, ConfigParser should not be case sensitive. Was there a problem loading the config file that motivated you to make these changes in the beginning?

I'll allow the integration of ChatGPT code in our codebase if you can explain to me what it does line by line.

fthouin commented on src/main.py:
```diff
@@ -31,7 +31,7 @@
 from samples.drivers.exemple_image_generation import beam_image_gen
 from drivers.Instruments import load_instruments
-from GUI.BeamExplorer import BeamExplorer
+from GUI.BeamExplorer import BeamExplorer, BeamWidget
```
`BeamWidget` is not called in the main. Why import it?

SDaneau replied:
Done, removed it.

fthouin commented on src/main.py:
```diff
@@ -540,13 +557,23 @@ def fitChirpMeasurement(self):
             poly_eq = " + ".join(f"c{i}" if i == 0 else f"c{i} * x^{i}" for i in range(len(coeffs)))
             lines = [f"Equation: {poly_eq}", ""] + [f"c{i} = {v:.2f} {'fs^2' if i == 0 else f'fs^{i+2}'}" for i, v in enumerate(coeffs)]
             self.chirp_coeff.setText('\n'.join(lines))
+            self.coeffs = np.array(np.concatenate(([0, 0], coeffs)))
```
Is my understanding correct? The way this is set creates an attribute .coeffs to main. If that is indeed the case, could you find a less generic name for this attribute like `last_temp_fit_coeffs`? This would help us better understand what this attribute is for, especially if it is part of a large thread like `main`

SDaneau replied:
Done, changed for self.last_temp_fit_coeffs

fthouin commented on src/measurements/CalibrationClasses.py:
```diff
@@ -311,11 +312,12 @@ class ChirpCalibrationMeasurement(QtCore.QThread):
-    def __init__(self,devices, background, grating_period, beam_, compression_carrier_wavelength, chirp_step, chirp_max, chirp_min,beam, demo=False):
+    def __init__(self,devices, background, grating_period, compression_carrier_wavelength, chirp_step, chirp_max, chirp_min, beam_name, beam, spectral_calibration=None, demo=False):
         '''
          Initializes the semporal beam calibration measurement
```
Not all inputs are described in the docstring. Please add the missing qualifiers

SDaneau replied:
Done

fthouin commented on src/measurements/CalibrationClasses.py:
```diff
@@ -342,25 +342,24 @@ def __init__(self,devices, background, grating_period, beam_, compression_carrie
+        if spectral_calibration == None:
+            self.beam.set_pixelToWavelength(Polynomial(1e-9*np.array([compression_carrier_wavelength-100,1/10]))) # arbitrairy polynomial spectral calibration
+        else:
+            print('no spectral calibration')       
```
Please replace this with [a exception raising scheme](https://realpython.com/python-raise-exception/)

SDaneau replied:
I don't think it's necessary to use the try-raise scheme as if there is no spectral calibration, the if function will set an arbitrary spectral calibration. The else is useless here since the chirp scan cannot be perform without a spectral calibration... I modified the code to include only the if and printing the use of an arbitrary spectral calibration.

fthouin replied:
A print outside the log to specify an exception was used is bad practice. You don`t need to use `try`  but you need to raise at least a Warning in the log, ideally an exception. Please follow current logging nomenclature it you go the log way.

SDaneau replied:
Done, a warning in the log is printed

fthouin commented on src/measurements/CalibrationClasses.py:
```diff
@@ -342,25 +342,24 @@ def __init__(self,devices, background, grating_period, beam_, compression_carrie
+        self.beam_name = beam_name
+        self.beam = beam
+        if spectral_calibration == None:
+            self.beam.set_pixelToWavelength(Polynomial(1e-9*np.array([compression_carrier_wavelength-100,1/10]))) # arbitrairy polynomial spectral calibration
```
What is the logic here? If no spectral calibration is present, why not simply raise an exception and abort the temporal calibration?

SDaneau replied:
That's right. The last topic get your point. However, it will used an arbitrary spectral calibration instead of abording the temporal calibration.

fthouin replied:
Why perform a temporal calibration if there is no spectral calibration? I insist, if there is no point, it should raise an error.

fthouin commented on src/measurements/CalibrationClasses.py:
```diff
@@ -509,23 +509,32 @@ def fit_chirp_scan(self, wavelength_array, chirp_array, data, deg, carrier_wavel
         # Convert wavelengts to frequency [Hz]
         c = 3e8 # speed of light in m/s
         freqs_values = c / (np.array(wavelength_values) * 1e-9) # Hz
+        omega_values = 2*np.pi*c / (np.array(wavelength_values) * 1e-9) # rad Hz
```
Please use the methods in `colbertoutils.py` for unit conversion

SDaneau replied:
Done

fthouin commented on src/measurements/CalibrationClasses.py:
```diff
@@ -311,11 +312,12 @@ class ChirpCalibrationMeasurement(QtCore.QThread):
     sendProgress = QtCore.pyqtSignal(float)
 
-    def __init__(self,devices, background, grating_period, beam_, compression_carrier_wavelength, chirp_step, chirp_max, chirp_min,beam, demo=False):
+    def __init__(self,devices, background, grating_period, compression_carrier_wavelength, chirp_step, chirp_max, chirp_min, beam_name, beam, spectral_calibration=None, demo=False):
```
Please add an entry in the manual describing the use and inner workings of the feature you implemented.

SDaneau replied:
Done

fthouin reviewed (changes requested):
Thanks a lot for this important contribution. You will find my review in the comments below.

Additionally, I get the following error when trying to run this on my virtual machine. 

```
  File "Z:\colberto\src\main.py", line 746, in <module>
    window = MainInterface()
  File "Z:\colberto\src\main.py", line 198, in __init__
    self.spec_length = self.devices['spectrometer'].get_num_pixel()
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'SpectrometerDemo' object has no attribute 'get_num_pixel'
```
when trying to run it on my machine.

SDaneau commented:
> Thanks a lot for this important contribution. You will find my review in the comments below.
> 
> Additionally, I get the following error when trying to run this on my virtual machine.
> 
> ```
>   File "Z:\colberto\src\main.py", line 746, in <module>
>     window = MainInterface()
>   File "Z:\colberto\src\main.py", line 198, in __init__
>     self.spec_length = self.devices['spectrometer'].get_num_pixel()
>                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'SpectrometerDemo' object has no attribute 'get_num_pixel'
> ```
> 
> when trying to run it on my machine.


@fthouin, added the attribute in the spectrometer demo

SDaneau committed:
get number of pixel in demo mode

SDaneau committed:
Hard coded calibration

I've put those numbers in the config_UdeM file

SDaneau committed:
Remove BeamWidget

Is not used

SDaneau committed:
Change self.coeffs name

for self.last_temp_fit_coeffs

SDaneau committed:
Add docstrings

SDaneau committed:
Arbitrary spectral calibration

SDaneau committed:
Use Colbert Outils

Cleaned the code for the chirp scan fit

SDaneau committed:
Calibration in instrument.py

SDaneau committed:
Update manual

SDaneau committed:
Update Instruments.py

fthouin commented:
> Did the piece of glass test to increase the GDD. The change of GDD for 2cm of UVFS should be around 700fs^2. I obtained a change of roughly 500 fs^2 which is close but still different from what is expected. I've put the glass piece before the SLM, just after the DOE. Maybe it's too far from the SLM, so it also change a bit the beam path. I'll bring the UVFS glass close to the SHG crystal so it's gonna minimize the change in beam aligment.

I think it is important we figure this out as it can be a symptom of miscalibration or incorrect phase setting. For reference, see a slide from a presentation I did while building the previous version of Colbert.
<img width="852" height="640" alt="image" src="https://github.com/user-attachments/assets/0bac1f80-b4a9-4807-aa9e-51e8965113e0" />

SDaneau commented:
> > Did the piece of glass test to increase the GDD. The change of GDD for 2cm of UVFS should be around 700fs^2. I obtained a change of roughly 500 fs^2 which is close but still different from what is expected. I've put the glass piece before the SLM, just after the DOE. Maybe it's too far from the SLM, so it also change a bit the beam path. I'll bring the UVFS glass close to the SHG crystal so it's gonna minimize the change in beam aligment.
> 
> I think it is important we figure this out as it can be a symptom of miscalibration or incorrect phase setting. For reference, see a slide from a presentation I did while building the previous version of Colbert. <img alt="image" width="852" height="640" src="https://private-user-images.githubusercontent.com/62530052/510895868-0bac1f80-b4a9-4807-aa9e-51e8965113e0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI1Mjk3MzcsIm5iZiI6MTc2MjUyOTQzNywicGF0aCI6Ii82MjUzMDA1Mi81MTA4OTU4NjgtMGJhYzFmODAtYjRhOS00ODA3LWFhOWUtNTFlODk2NTExM2UwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA3VDE1MzAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwMDY2NjA2YTYyMjcxY2NkYjVmZWQ1NWU2MjhjYzQzYWFhN2MzMjZjOTNhOWFmYTgyMTkzN2RiMTZhZjg0NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LevCN_uLpItxHU8rmGqyGfAP4T-bpP7wjpBSVhLRYcM">

Yes, it's important. However, I did the test again and I don't get a nice shift over all the frequency. I'm suppose to have a change of +720fs^2 but I'm getting -660fs^2 for some wavelength and +940fs^2 for other... looks weird.  I'm gonna try again, the spectral calibration is good, so I think we are facing an incorrect phase setting... here the GVD of UVFS.

<img width="800" height="584" alt="image" src="https://github.com/user-attachments/assets/0f3ab6da-01b9-4477-9a42-62d1e892e13c" />

SDaneau committed:
Requested modifications

SDaneau committed:
Include the prefactor of the Taylor serie

SDaneau committed:
Final details

SDaneau committed:
Remove print

SDaneau committed:
Update Slm_Meadowlark_optics.py

SDaneau committed:
Update main.py

fthouin commented:
Great work from Simon [demonstrated accurate compression using chirp scan](https://openproject.silvascience.org/projects/colberto/work_packages/849/activity#comment-10738). @SDaneau could you pull dev into this branch to fix the conflict with main? We can merge afterwards.

SDaneau commented:
> Great work from Simon [demonstrated accurate compression using chirp scan](https://openproject.silvascience.org/projects/colberto/work_packages/849/activity#comment-10738). @SDaneau could you pull dev into this branch to fix the conflict with main? We can merge afterwards.

I did merge dev into my branch and the only conflict was in the main, resolved by combining the two. The difference was coming from the definition that close all windows at the same time.

fthouin commented on output_image.png:
Can you remove this?

fthouin commented on src/collcyl_correctoffset_bk7.txt:
Can you remove this file from the commit as it is was only for testing purposes?

fthouin commented on src/collcyl_correctoffset_nobk7.txt:
same here

fthouin commented on src/drivers/Slm_Meadowlark_optics.py:
Please remove these changes from the commit. There shouldn't be any changes in this file right?

fthouin reviewed (changes requested):
Last little things to change!!!

SDaneau committed:
Merge branch 'dev' into Simon_IterativeCompression

SDaneau committed:
Last changes

SDaneau committed:
Update Slm_Meadowlark_optics.py

SDaneau committed:
Delete output_image.png
